Error: Minified React error #185 displays when attempting to log into Graylog

<!--- Provide a general summary of the issue in the Title above -->
**Note: This problem only happens when running a production version of Graylog from build artifacts (tar/jar etc). It does not happen in the Graylog development environment.**

## Expected Behavior
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->
The Graylog login page should show after server startup when attempting to access Graylog in the web browser.

## Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->
The following error page displays instead.
![image](https://user-images.githubusercontent.com/3423655/163272301-3725ce8d-7a9e-4792-a771-a8147ff62c01.png)

<summary>
Full stacktrace
<details>
vendor.42872bd04f88743491d7.js:2 Error: Minified React error #185; visit https://reactjs.org/docs/error-decoder.html?invariant=185 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at nc (vendor.42872bd04f88743491d7.js:2:836854)
    at Object.enqueueSetState (vendor.42872bd04f88743491d7.js:2:795588)
    at t.y.setState (vendor.42872bd04f88743491d7.js:2:951729)
    at vendor.42872bd04f88743491d7.js:2:936552
    at r (history.js:155:24)
    at history.js:173:23
    at Array.forEach (<anonymous>)
    at Object.notifyListeners (history.js:172:15)
    at P (history.js:288:23)
    at history.js:402:11
    at Object.confirmTransitionTo (history.js:145:7)
    at Object.replace (history.js:385:23)
    at t._redirect (StartPage.jsx:54:5)
    at t._redirectToStartpage (StartPage.jsx:63:14)
    at t.componentDidUpdate (StartPage.jsx:45:12)
    at us (vendor.42872bd04f88743491d7.js:2:830011)
    at gc (vendor.42872bd04f88743491d7.js:2:847554)
    at exports.unstable_runWithPriority (vendor.42872bd04f88743491d7.js:2:867223)
    at Fi (vendor.42872bd04f88743491d7.js:2:791632)
    at vc (vendor.42872bd04f88743491d7.js:2:844220)
    at sc (vendor.42872bd04f88743491d7.js:2:840374)
    at vendor.42872bd04f88743491d7.js:2:791923
    at exports.unstable_runWithPriority (vendor.42872bd04f88743491d7.js:2:867223)
    at Fi (vendor.42872bd04f88743491d7.js:2:791632)
    at Vi (vendor.42872bd04f88743491d7.js:2:791868)
    at Gi (vendor.42872bd04f88743491d7.js:2:791803)
    at nc (vendor.42872bd04f88743491d7.js:2:836960)
    at Object.enqueueSetState (vendor.42872bd04f88743491d7.js:2:795588)
    at t.y.setState (vendor.42872bd04f88743491d7.js:2:951729)
    at t.onGettingStartedUpdate (StartPage.jsx:50:10)
    at i.n (vendor.42872bd04f88743491d7.js:2:959228)
    at i.emit (vendor.42872bd04f88743491d7.js:2:56868)
    at p.trigger (vendor.42872bd04f88743491d7.js:2:960109)
    at GettingStartedStore.js:58:18
is @ vendor.42872bd04f88743491d7.js:2
</details>
</summary>


## Possible Solution
<!--- Not obligatory, but suggest a fix/reason for the bug, -->
<!--- or ideas how to implement the addition or change -->

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
As mentioned above, this problem only happens when running a production version of Graylog from build artifacts (tar/jar etc). It does not happen in the Graylog development environment.

From investigating with @mikedklein, the problem seems to exclusively occur when running a build of Graylog from the `master` branch. It does **not** happen when running from the `4.3` branch. It appears that something has been introduced to `master` that is triggering the issue. I have not had time to investigate what yet.

You can reproduce the issue with the latest `master` branch build from Jenkins: https://jenkins.ci.torch.sh/blue/organizations/jenkins/Graylog-Snapshots%2Fgraylog2-server/detail/master/2118/artifacts

Download a snapshot build tar, start up Graylog, and the issue should happen. 

## Additional Details
@mikedklein mentioned that the error is not displayed if you don't currently have an active Graylog session. (the login page displays instead), but the error appears as you attempt to access some other pages.

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->

* Graylog Version: 4.4.0 SNAPSHOT (master)
* Java Version: 1.8.0_181
* Elasticsearch Version: Opensearch 1.1
* MongoDB Version: 
* Operating System: Mac OS Monererry 12.3.1
* Browser version:

--- a/graylog2-web-interface/src/routing/Routes.test.ts
+++ b/graylog2-web-interface/src/routing/Routes.test.ts
@@ -77,4 +77,38 @@ describe('Routes', () => {
       expect(uri.hasQuery('interval', 'hour')).toBeTruthy();
     });
   });
+
+  describe('with `/` prefix', () => {
+    beforeAll(() => {
+      jest.resetModules();
+
+      window.appConfig = {
+        gl2AppPathPrefix: '/',
+      } as AppConfigs;
+
+      Routes = jest.requireActual('./Routes').default;
+    });
+
+    it('returns a route from constant', () => {
+      expect(Routes.SEARCH).toMatch('/search');
+    });
+
+    it('returns correct route for getting started page', () => {
+      expect(Routes.GETTING_STARTED).toMatch('/gettingstarted');
+    });
+
+    it('returns a route from function', () => {
+      expect(Routes.node('id')).toMatch('/system/nodes/id');
+    });
+
+    it('routes contain query parameters', () => {
+      const uri = URI(Routes.search('', { rangetype: 'relative', relative: 300 }, 'hour'));
+
+      expect(uri.path()).toMatch('/search');
+      expect(uri.hasQuery('q', '')).toBeTruthy();
+      expect(uri.hasQuery('rangetype', 'relative')).toBeTruthy();
+      expect(uri.hasQuery('relative', '300')).toBeTruthy();
+      expect(uri.hasQuery('interval', 'hour')).toBeTruthy();
+    });
+  });
 });

--- a/graylog2-web-interface/src/routing/Routes.ts
+++ b/graylog2-web-interface/src/routing/Routes.ts
@@ -254,19 +254,28 @@ const Routes = {
   filtered_metrics: (nodeId: string, filter: string) => `${Routes.SYSTEM.METRICS(nodeId)}?filter=${filter}`,
 } as const;
 
+const prefixUrlWithoutHostname = (url: string, prefix: string) => {
+  const uri = new URI(url);
+
+  return uri.directory(`${prefix}/${uri.directory()}`)
+    .normalizePath()
+    .resource();
+};
+
 const qualifyUrls = (routes, appPrefix): typeof routes => {
   const qualifiedRoutes = {};
 
   Object.keys(routes).forEach((routeName) => {
     switch (typeof routes[routeName]) {
       case 'string':
-        qualifiedRoutes[routeName] = new URI(`${appPrefix}/${routes[routeName]}`).normalizePath().resource();
+        qualifiedRoutes[routeName] = prefixUrlWithoutHostname(routes[routeName], appPrefix);
+
         break;
       case 'function':
         qualifiedRoutes[routeName] = (...params) => {
           const result = routes[routeName](...params);
 
-          return new URI(`${appPrefix}/${result}`).normalizePath().resource();
+          return prefixUrlWithoutHostname(result, appPrefix);
         };
 
         break;

--- a/graylog2-web-interface/src/routing/Routes.ts
+++ b/graylog2-web-interface/src/routing/Routes.ts
@@ -262,36 +262,33 @@ const prefixUrlWithoutHostname = (url: string, prefix: string) => {
     .resource();
 };
 
-const qualifyUrls = (routes, appPrefix): typeof routes => {
+type RouteFunction<P extends Array<any>> = (...args: P) => string;
+type RouteMapEntry = string | RouteFunction<any> | RouteMap;
+type RouteMap = { [routeName: string]: RouteMapEntry };
+
+const isLiteralRoute = (entry: RouteMapEntry): entry is string => (typeof entry === 'string');
+const isRouteFunction = (entry: RouteMapEntry): entry is RouteFunction<any> => (typeof entry === 'function');
+
+const qualifyUrls = <R extends RouteMap>(routes: R, appPrefix: string): R => {
   if (appPrefix === '/') {
     return routes;
   }
 
-  const qualifiedRoutes = {};
-
-  Object.keys(routes).forEach((routeName) => {
-    switch (typeof routes[routeName]) {
-      case 'string':
-        qualifiedRoutes[routeName] = prefixUrlWithoutHostname(routes[routeName], appPrefix);
-
-        break;
-      case 'function':
-        qualifiedRoutes[routeName] = (...params) => {
-          const result = routes[routeName](...params);
-
-          return prefixUrlWithoutHostname(result, appPrefix);
-        };
-
-        break;
-      case 'object':
-        qualifiedRoutes[routeName] = qualifyUrls(routes[routeName], appPrefix);
-        break;
-      default:
-        break;
+  return Object.fromEntries(Object.entries(routes).map(([routeName, routeValue]) => {
+    if (isLiteralRoute(routeValue)) {
+      return [routeName, prefixUrlWithoutHostname(routeValue, appPrefix)];
     }
-  });
 
-  return qualifiedRoutes;
+    if (isRouteFunction(routeValue)) {
+      return [routeName, (...params: Parameters<typeof routeValue>) => {
+        const result = routeValue(...params);
+
+        return prefixUrlWithoutHostname(result, appPrefix);
+      }];
+    }
+
+    return [routeName, qualifyUrls(routeValue, appPrefix)];
+  }));
 };
 
 const qualifiedRoutes: typeof Routes = AppConfig.gl2AppPathPrefix() ? qualifyUrls(Routes, AppConfig.gl2AppPathPrefix()) : Routes;
@@ -314,7 +311,7 @@ const unqualified = Routes;
  * <LinkContainer to={Routes.pluginRoutes('SYSTEM_PIPELINES_PIPELINEID')(123)}>...</LinkContainer>
  *
  */
-const pluginRoute = (routeKey, throwError = true) => {
+const pluginRoute = (routeKey: string, throwError: boolean = true) => {
   const pluginRoutes = {};
 
   PluginStore.exports('routes').forEach((route) => {
@@ -349,7 +346,7 @@ const pluginRoute = (routeKey, throwError = true) => {
   return route;
 };
 
-const getPluginRoute = (routeKey) => pluginRoute(routeKey, false);
+const getPluginRoute = (routeKey: string) => pluginRoute(routeKey, false);
 
 /**
  * Exported constants for using strings to check if a plugin is registered by it's description.

--- a/graylog2-web-interface/src/routing/Routes.ts
+++ b/graylog2-web-interface/src/routing/Routes.ts
@@ -263,6 +263,10 @@ const prefixUrlWithoutHostname = (url: string, prefix: string) => {
 };
 
 const qualifyUrls = (routes, appPrefix): typeof routes => {
+  if (appPrefix === '/') {
+    return routes;
+  }
+
   const qualifiedRoutes = {};
 
   Object.keys(routes).forEach((routeName) => {

